object_tracker/object_tracker.go

```diff
@@ -48,57 +71,108 @@ func newTracker(ctx context.Context, deps resource.Dependencies, conf resource.C
 		return nil, err
 	}
-
 	// Default value for frequency = 10Hz
 	if t.frequency == 0 {
 		t.frequency = DefaultMaxFrequency
 	}
 
-	// Populate the first set of 2 detections to start us off
-	stream, err := t.cam.Stream(ctx, nil)
+	cancelableCtx, cancel := context.WithCancel(context.Background())
+	t.cancelFunc = cancel
+	t.cancelContext = cancelableCtx
+
+	// Do the first pass to populate the first set of 2 detections.
+	starterDets := make([][]objdet.Detection, 2)
+	stream, err := t.cam.Stream(t.cancelContext, nil)
 	if err != nil {
 		return nil, err
 	}
 	for i := 0; i < 2; i++ {
-		img, _, err := stream.Next(ctx)
+		img, _, err := stream.Next(t.cancelContext)
 		if err != nil {
 			return nil, err
 		}
-		detections, err := t.detector.Detections(ctx, img, nil)
+		detections, err := t.detector.Detections(t.cancelContext, img, nil)
 		if err != nil {
 			return nil, err
 		}
-		t.oldDetections[i] = detections
+		starterDets[i] = detections
 	}
-
-	filteredOld := FilterDetections(t.chosenLabels, t.oldDetections[0])
-	filteredNew := FilterDetections(t.chosenLabels, t.oldDetections[1])
-
-	// Rename from scratch
-	// The strategy is to rename the (n-1) detections and then match the (n) dets to those
-	// When a (n) det matches a (n-1) det, it copies its label
+	filteredOld := FilterDetections(t.chosenLabels, starterDets[0])
+	filteredNew := FilterDetections(t.chosenLabels, starterDets[1])
+	// Rename (from scratch)
 	renamedOld := make([]objdet.Detection, 0, len(filteredOld))
 	for _, det := range filteredOld {
 		newDet := t.RenameFirstTime(det)
 		renamedOld = append(renamedOld, newDet)
 	}
-
 	// Build and solve cost matrix via Munkres' method
 	matchMtx := t.BuildMatchingMatrix(renamedOld, filteredNew)
 	HA, err := hg.NewHungarianAlgorithm(matchMtx)
 	if err != nil {
 		return nil, err
 	}
 	matches := HA.Execute()
-
 	// Rename from temporal matches. New det copies old det's label
 	renamedNew := t.RenameFromMatches(matches, renamedOld, filteredNew)
-	t.oldDetections[0] = renamedOld
-	t.oldDetections[1] = renamedNew
+	t.oldDetections.Store(&[2][]objdet.Detection{renamedOld, renamedNew})
+
+	t.activeBackgroundWorkers.Add(1)
+	viamutils.ManagedGo(func() {
+		t.run(stream, t.cancelContext)
+	}, func() {
+		t.cancelFunc()
+		stream.Close(t.cancelContext)
+		t.activeBackgroundWorkers.Done()
+	})
 
 	return t, nil
 }
 
+// run is a (cancelable) infinite loop that takes new detections from the camera and compares them to
+// the most recently seen detections. Matching detections are linked via matching labels.
+func (t *myTracker) run(stream gostream.VideoStream, cancelableCtx context.Context) {
+	for {
+		select {
+		case <-cancelableCtx.Done():
+			return
+		default:
+			streamCtx := context.Background()
+			start := time.Now()
+			// Load up the old detections
+			namedOld := t.oldDetections.Load()[1]
+
+			// Take fresh detections from fresh image
+			img, _, err := stream.Next(streamCtx)
+			if err != nil {
+				return
```

Comment:
Also, if your err != nil, you really should still at least print what the error is (to the logger)